### PR TITLE
Make FlowMetric mostly immutable

### PR DESCRIPTION
**Pull request by mattdurham**
#### PR Description

Make FlowMetrics semi-immutable, before a flowmetric could be easily changed. There could be more memory alloc/gced due to LabelsCopy() but that is likely unavoidable.

### CODE REVIEW

File: component/metrics/mutate/mutate.go [hunk 1]
```diff
@@ -101,19 +102,21 @@ func (c *Component) Update(args component.Arguments) error {
 // series. This is a blocker for releasing a production-grade  of the metrics.mutate
 // component.
 func (c *Component) Receive(ts int64, metricArr []*metrics.FlowMetric) {
-	app := c.appendable.Appender(context.Background())
+	c.mut.Lock()
+	defer c.mut.Unlock()
```
Comment (rfratto) [inline component/metrics/mutate/mutate.go:106]: This should probably be a RLock/RUnlock since many things might be sending metrics at the same time

File: component/metrics/mutate/mutate.go [hunk 2]
```diff
@@ -101,19 +102,21 @@ func (c *Component) Update(args component.Arguments) error {
...
-		if m.Labels == nil {
+		newLabels := relabel.Process(m.LabelsCopy(), c.mrc...)
+		if newLabels == nil {
 			continue
 		}
-		_, err := app.Append(storage.SeriesRef(m.GlobalRefID), m.Labels, ts, m.Value)
-		if err != nil {
-			level.Error(c.opts.Logger).Log("msg", "failed to forward sample from metrics.mutate component", "err", err, "componentID", c.opts.ID)
-		}
+		relabelledMetrics = append(relabelledMetrics, metrics.NewFlowMetric(metrics.GlobalRefMapping.GetOrAddGlobalRefID(newLabels), newLabels, m.Value()))
```
Comment (rfratto) [inline component/metrics/mutate/mutate.go:114]: nit: kind of a long line, can we assign the new metric to a variable rather than constructing it in the append?

File: component/metrics/receiver.go [hunk 3]
```diff
@@ -13,9 +14,35 @@ type Receiver struct {
...
 func (r Receiver) RiverCapsule() {}
 
-// FlowMetric is a wrapper around a single metric without the timestamp
+// FlowMetric is a wrapper around a single metric without the timestamp, FlowMetric is *mostly* immutable
+// the globalrefid can only be set if it is 0
 type FlowMetric struct {
-	GlobalRefID uint64
-	Labels      labels.Labels
-	Value       float64
+	globalRefID *atomic.Uint64
```
Comment (rfratto) [inline component/metrics/receiver.go:20]: This represents a single in-flight sample, right? I'm not sure I understand when you would change this in-flight.

File: component/metrics/receiver.go [hunk 4]
```diff
@@ -13,9 +14,35 @@ type Receiver struct {
...
+		value:       value,
+	}
+}
+
+// GlobalRefID Retrieves the GlobalRefID
+func (fw *FlowMetric) GlobalRefID() uint64 { return fw.globalRefID.Load() }
+
+// SetGlobalRefID will only set the GlobalRefID if it is 0
+func (fw *FlowMetric) SetGlobalRefID(globalRefID uint64) {
+	fw.globalRefID.CAS(0, globalRefID)
```
Comment (rfratto) [inline component/metrics/receiver.go:39]: The CAS only works if the previous ref ID was 0, which it almost never will be.

File: component/metrics/remotewrite/remote_write.go [hunk 5]
```diff
@@ -174,15 +174,15 @@ func (c *Component) Receive(ts int64, metricArr []*metrics.FlowMetric) {
...
-			globalID := metrics.GlobalRefMapping.GetOrAddGlobalRefID(m.Labels)
-			m.GlobalRefID = globalID
+		if m.GlobalRefID() == 0 {
+			globalID := metrics.GlobalRefMapping.GetOrAddGlobalRefID(m.LabelsCopy())
+			m.SetGlobalRefID(globalID)
 		}
-		localID := metrics.GlobalRefMapping.GetLocalRefID(c.opts.ID, m.GlobalRefID)
-		newLocal, err := app.Append(storage.SeriesRef(localID), m.Labels, ts, m.Value)
+		localID := metrics.GlobalRefMapping.GetLocalRefID(c.opts.ID, m.GlobalRefID())
+		newLocal, err := app.Append(storage.SeriesRef(localID), m.LabelsCopy(), ts, m.Value())
```
Comment (rfratto) [inline component/metrics/remotewrite/remote_write.go:182]: This seems a bit like a code smell. If the intent is that senders should generate their own ref IDs, then we should rely on that here and never mutate the in-flight samples. 

Plus, if the metrics are in-flight and they were generated ad-hoc from the sender, it's not going to get the updated ref ID as it is. 

I think this should just be 

```go
localID := metrics.GlobalRefMapping.GetOrAddGlobalRefID(m.Labels()) 
```

File: component/metrics/receiver.go [hunk 6]
```diff
@@ -13,9 +14,35 @@ type Receiver struct {
...
+}
+
+// Value returns the value
+func (fw *FlowMetric) Value() float64 { return fw.value }
+
+// LabelsCopy returns a copy of the labels structure
+func (fw *FlowMetric) LabelsCopy() labels.Labels {
+	return fw.labels.Copy()
 }
```
Comment (rfratto) [inline component/metrics/receiver.go:48]: There should be a read-only method for labels. Making a copy every time we want to read will be expensive.

**Review by rfratto — COMMENTED** [review]
This is shaping up ok, but I'm a bit concerned about the changes as they are.

File: component/metrics/receiver.go (hunk 3 above)
Comment (mattdurham) [reply component/metrics/receiver.go:20]: It's an escape hatch to ensure items have a global id set for the remote_write. It might not be needed, but ensures any metrics have an id assigned.

File: component/metrics/receiver.go (hunk 4 above)
Comment (mattdurham) [reply component/metrics/receiver.go:39]: That is by design, if we don't want the double checking then we can make it immutable.

Comment (rfratto) [reply component/metrics/receiver.go:39]: Sorry, I missed the doc comment. The API here feels a little hard to follow, I think we should spend some time investing in making it easier to reason about.

File: component/metrics/receiver.go (hunk 6 above)
Comment (mattdurham) [reply component/metrics/receiver.go:48]: When that use case arises, my plan was to add features on the FlowMetric itself. 
* ContainsLabel
* GetValue
etc

Comment (rfratto) [reply component/metrics/receiver.go:48]: If you're firm on copying the labels, can we write benchmarks for metrics.remote_write's receiver? 

We need to be cautious of changes that adds more overhead.

Comment (mattdurham) [reply component/metrics/receiver.go:48]: When benching `GlobalRefMapping.GetGlobalRefIDByLabels(fms[j].labels)` vs `GlobalRefMapping.GetGlobalRefIDByLabels(fms[j].LabelsCopy())` the second is roughly 57% slower.

Comment (mattdurham) [reply component/metrics/receiver.go:48]: Fiddling around to where the GlobalRefMapping grabs the underlying labels since it is in the same package and is safe to use in this specific instance.

**Comment by mattdurham** [comment]
Simplified the flowmetric and the surface area of the GlobalRefMapping, GlobalRefMapping now uses the fact its in the same space as the FlowMetric to access the internal labels. It knows that it is safe to do this. Was able to remove several function calls or unexport them for smaller surface area. Finally made the globalrefid its own type so it wasn't easy to confuse a local id for a global.

File: component/metrics/remotewrite/remote_write.go (hunk 5 above)
Comment (mattdurham) [reply component/metrics/remotewrite/remote_write.go]: Simplified this code path

File: component/common/appendable/appendable.go [hunk 7]
```diff
@@ -69,25 +63,25 @@ func (app *flowAppender) Append(ref storage.SeriesRef, l labels.Labels, t int64,
...
 
-func (app *flowAppender) AppendExemplar(ref storage.SeriesRef, l labels.Labels, e exemplar.Exemplar) (storage.SeriesRef, error) {
+func (app *flowAppender) AppendExemplar(
+	ref storage.SeriesRef,
+	l labels.Labels,
+	e exemplar.Exemplar,
+) (storage.SeriesRef, error) {
+
 	return 0, nil
 }
```
Comment (rfratto) [inline component/common/appendable/appendable.go:86]: Can you add a TODO here? This is something we'll want to implement before release.

File: component/metrics/globalrefmap.go [hunk 8]
```diff
@@ -89,8 +89,8 @@ func (g *GlobalRefMap) GetOrAddGlobalRefID(l labels.Labels) uint64 {
 	return g.globalRefID
 }
 
-// GetGlobalRefID returns the global refid for a component local combo, or 0 if not found
-func (g *GlobalRefMap) GetGlobalRefID(componentID string, localRefID uint64) uint64 {
+// GetGlobalRefIDForComponent returns the global refid for a component local combo, or 0 if not found
+func (g *GlobalRefMap) GetGlobalRefIDForComponent(componentID string, localRefID uint64) RefID {
```
Comment (rfratto) [inline component/metrics/globalrefmap.go:93]: I just noticed that `g.mappings` never gets cleaned up, even after a component goes away, so mappings will leak. Can you add a TODO somewhere to fix the leak?

File: component/metrics/receiver.go [hunk 9]
```diff
@@ -13,9 +13,37 @@ type Receiver struct {
 // RiverCapsule marks receivers as a capsule.
 func (r Receiver) RiverCapsule() {}
 
-// FlowMetric is a wrapper around a single metric without the timestamp
+// FlowMetric is a wrapper around a single metric without the timestamp, FlowMetric is *mostly* immutable
+// the globalrefid can only be set if it is 0
```
Comment (rfratto) [inline component/metrics/receiver.go:17]: Is the "mostly" comment stale now?

File: component/metrics/remotewrite/remote_write.go [hunk 10]
```diff
@@ -173,16 +173,11 @@ func (c *Component) Update(newConfig component.Arguments) error {
...
 	for _, m := range metricArr {
-		// TODO this should all be simplified into one call
-		if m.GlobalRefID == 0 {
-			globalID := metrics.GlobalRefMapping.GetOrAddGlobalRefID(m.Labels)
-			m.GlobalRefID = globalID
-		}
-		localID := metrics.GlobalRefMapping.GetLocalRefID(c.opts.ID, m.GlobalRefID)
-		newLocal, err := app.Append(storage.SeriesRef(localID), m.Labels, ts, m.Value)
+		localID := metrics.GlobalRefMapping.GetLocalRefID(c.opts.ID, m.GlobalRefID())
+		newLocal, err := app.Append(storage.SeriesRef(localID), m.LabelsCopy(), ts, m.Value())
```
Comment (rfratto) [inline component/metrics/remotewrite/remote_write.go:177]: I'm still concerned about this making a copy of the labels for every single sample.

Comment (mattdurham) [reply component/metrics/remotewrite/remote_write.go:177]: If we ever expose relabel_configs for remote_write then they could mutate the labels.

File: component/metrics/globalrefmap.go (hunk 8 above)
Comment (mattdurham) [reply component/metrics/globalrefmap.go:93]: Fixed the leak.

File: component/common/appendable/appendable.go [hunk 11]
```diff
@@ -69,34 +63,35 @@ func (app *flowAppender) Append(ref storage.SeriesRef, l labels.Labels, t int64,
...
+	e exemplar.Exemplar,
+) (storage.SeriesRef, error) {
+	// TODO actually implement this before flow production.
 	return 0, nil
 }
 
 func (app *flowAppender) Commit() error {
 	for _, r := range app.receivers {
 		for ts, metrics := range app.buffer {
-			if r == nil || r.Receive == nil {
```
Comment (rfratto) [inline component/common/appendable/appendable.go:97]: Why are we removing this check? IIRC this was added at some point to prevent a panic.

**Review by rfratto — CHANGES_REQUESTED** [review]
I'd prefer for us to not merge this in its current state. 

Since we've discovered how much time we're spending in the GC, I'm hard against us copying labels for every received sample, which will exacerbate the problem. We should be matching Prometheus' behavior for labels: do copy-on-write prior to when label modifications are going to be done, not copy-on-read. 

I'd suggest building a Docker image for this and running it in our dev clusters for a while to see what impact it has on CPU. If I'm wrong, I'll retract my concerns, but I'm fairly certain this will have a significant performance hit.

File: component/common/appendable/appendable.go (hunk 11 above)
Comment (mattdurham) [reply component/common/appendable/appendable.go:97]: I was going to ensure that never happens at the receiver level but will add that back in until I make that change in a separate PR.
